### ISSUE TITLE
Upload script never finishes after some types of network errors

### ISSUE THREAD

**Issue by ToeBee** [mapillary/mapillary_tools#194]
Sometimes I see an HTTP error response during uploads and either the picture actually did make it to the server or the script tries again because the sequence ends up with the correct number of pictures on the server and the upload script exits normally.

However sometimes (I think if there is an error establishing the TCP connection instead of the server returning an HTTP error code) the script will end up with an unhandled exception which results in one of the upload threads going zombie and never finishing. This means the script never sends the `DONE` request to the server when it finishes uploading the images and also does not exit because one of the threads is still running. So I have to kill it and then go to the legacy site to publish the sequence.

Here is the console output from when this happened earlier today. It does seem like all the images did get uploaded successfully though.
```Uploading sequence d316ccee-ccc7-4cad-983d-91be9127600f.
Uploading 4267 images with valid exif tags (Skipping 0) ...
Exception in thread Thread-4:--------------------------------] 41.7% ... 2488 images left.
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/home/toby/git/mapillary_tools/python/lib/uploader.py", line 44, in run
    upload_file(filepath, **self.params)
  File "/home/toby/git/mapillary_tools/python/lib/uploader.py", line 255, in upload_file
    response = urllib2.urlopen(request)
  File "/usr/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/usr/lib/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 1241, in https_open
    context=self._context)
  File "/usr/lib/python2.7/urllib2.py", line 1201, in do_open
    r = h.getresponse(buffering=True)
  File "/usr/lib/python2.7/httplib.py", line 1136, in getresponse
    response.begin()
  File "/usr/lib/python2.7/httplib.py", line 453, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python2.7/httplib.py", line 417, in _read_status
    raise BadStatusLine(line)
BadStatusLine: ''

^C===========================================================] 100.0% ... 0 images left...
BREAK: Stopping upload.
```

**Comment by jernejaMislej** [comment]
hi,
i had a look at this and the most likely issue is with sending the DONE file, which is equivalent to approving publication of uploaded images. 
Currently the DONE file is send in the end after all the images are uploaded, for each sequence.
I increased the max attempts for the DONE file and implemented a `post_process` command which includes pushing uploading images to be approved.
Generally if images were uploaded, but not approved, one can just run the `upload` command again and it will not re-upload images put just send the DONE file only. so if tools are just hanging there you can interrupt and run again. In some cases of bugs and also unknown issues images can get logged as if already approved for publication, but the DONE file is not successfully uploaded, in those cases the `upload` command will not do anything.
The `post_process` command will approve publication of all images in the specified import path regardless if logs state that it has already been done.
currently that is not the final solution but a temporary work around, as we are planning to improve this in the future releases
pushing images in the specified import path, ie uploading the DONE file will not result in any issues in any case, so you can run the `post_process` command with arg `--push_images` whenever you are not sure if the upload has been approved for publish or are missing uploaded images for a longer period of time
`post_process` command will be released with the next version, currently it is available at the v0.2.0 branch

**Comment by ToeBee** [comment]
Well this issue was opened against the old version of the upload script, before the complete rewrite. If I remember correctly the old version did not have functionality to resume a cancelled upload so rerunning the script did not work. Hence having to log into the legacy site and push the button.

I am not convinced that the problem is with sending the `DONE` file though. In my case, the network error indicated in the output happened in the middle of a 4,000+ image upload. Then it continued on, uploading several thousand more images without further incident. But then you think it failed to push a few bytes for the `DONE` request? I still think that some types of network exceptions break your error handling and result in a hung thread. I have seen this kind of thing happen in my own code (unrelated to Mapillary) before. However I will close this issue and follow along on #250 since it was opened against a more current version of the script.
